Use stages pipeline

dougbu opened the pull request:
- #11924
- change default-build.yml to use job.yml
  - remove unused parameters e.g. `poolName` and `matrix`
  - use `enableMicrobuild`, `enablePublishBuildArtifacts` etc. to simplify YAML scripts
- change ci.yml to use stages and post-build.yml
  - add use of publish-build-assets.yml and post-build.yml

dougbu committed:
Use stages pipeline
- #11924
- change ci.yml to use stages and post-build.yml
  - add use of publish-build-assets.yml and post-build.yml
  - publish in last build step of each job
    - pass more MSBuild properties into builds
  - use distinct `$(AssetManifestFileName)` values per job
    - set global property to override what's hard-coded in Publish.proj
- change codesign-xplat.yml to use empty.proj and normal Arcade signing and publication process
  - remove XPlatPackageSigner.proj
- change default-build.yml to use job.yml
  - remove unused parameters e.g. `matrix`, `poolName`, `variables`
  - use `enableMicrobuild` and `enablePublishTestResults` to eliminate duplicate build steps
- add .dll's and .exe's as files to sign w/ Microsoft400
  - add signcheck exclusions
- remove custom manifest generation i.e. the `GenerateBuildAssetManifest` target and related artifacts
  - use regular Arcade publish infrastructure
- update docker infrastructure to use same paths in and out of the container
- correct typo in build.sh

nits:
- upload logs in first artifact
- remove attempts to package non-existent VSIX
  - follow-up to 29cf7ecb80b1
- respect verbosity setting in build.sh
- add more information to Artifacts.md
- enable test signing in internal PRs

dougbu commented:
This seems to be getting close. But, job names may still be messed up i.e. job.yml seems to use `displayName` instead of `name` parameter.

Internal build: https://dev.azure.com/dnceng/internal/_build/results?buildId=305555&view=results

dougbu commented:
/cc @mmitche @riarenas

dougbu commented on .azure/pipelines/ci.yml:
```diff
@@ -22,6 +22,12 @@ variables:
+- name: _DotNetArtifactsCategory
+  value: .NETCORE
+- name: _DotNetValidationArtifactsCategory
+  value: .NETCOREVALIDATION # Reviwers: Should this be .NETCORE ?
```
???

wtgodbe replied:
I think this should match `_DotNetArtifactsCategory` - @riarenas can you confirm?

riarenas replied:
While you're testing the changes, it should match so that you see everything going to the feed you would expect it to go for the real thing. Once you're satsified with your testing, the variable should be removed.

dougbu replied:
1. Will change the value.
2. I left the `_DotNetValidationArtifactsCategory` setting in the Blazor repo. Does that matter?

riarenas replied:
Will change the value.
> I left the _DotNetValidationArtifactsCategory setting in the Blazor repo. Does that matter?
Only if you plan to publish to the tools validation channel for your branch. The publishing for that channel will go to a different feed, but I don't believe you want to publish to that channel anyways.

dougbu replied:
To make sure I have this right: As long as it's only my deleted branch that has a default channel mapping to the validation channel, it just doesn't matter?

riarenas replied:
Correct

dougbu commented on .azure/pipelines/ci.yml:
```diff
@@ -30,507 +36,534 @@ variables:
-      path: artifacts/installers/
-    - name: Windows_arm_Logs
-      path: artifacts/log/
-      publishOnError: true
```
Moving these was a nit. Best to get the logs uploaded first in case of a cancelation.

dougbu commented on .azure/pipelines/jobs/default-build.yml:
```diff
@@ -75,182 +65,158 @@ parameters:
-        ${{ if eq(parameters.isTestingJob, true) }}:
+- template: /eng/common/templates/job/job.yml
+  parameters:
+    name: ${{ coalesce(parameters.jobName, parameters.agentOs) }}
```
When I used `enablePublishBuildArtifacts`, it seemed `$(Agent.JobName)` matched `displayName` instead of `name`. Thus, I'm worried about connections made in codesign-xplat and the BAR publication step. Is this setup invalid?

wtgodbe commented:
Changes look reasonable, and the internal build seems good other than failing to publish to BAR:

> Artifact AssetManifests not found for build 305555. Please ensure you have published artifacts in any previous phases of the current build.

Maybe due to `_DotNetValidationArtifactsCategory` not matching `_DotNetArtifactsCategory`?

dougbu commented:
> Maybe due to _DotNetValidationArtifactsCategory not matching _DotNetArtifactsCategory?

I think this was due to using distinct artifact names for the various jobs pushing manifests.

dougbu commented:
New internal build is https://dev.azure.com/dnceng/internal/_build/results?buildId=306146

dougbu commented:
A test failed in https://dev.azure.com/dnceng/internal/_build/results?buildId=306146 Will need to wait a bit before we see how this is really working.

dougbu commented:
@riarenas @JohnTortugo @mmitche this is changing our artifact layout significantly. I _suspect_ I could merge all of the *_Packages_Signed artifacts with the Windows ones to create a single PackageArtifacts artifact (or perhaps duplicate the two) as I had to do w/ AssetManifests. But, am I just peeling an onion with multiple additional layers? (Less of a concern that I'm breaking our existing release pipeline of course -- that will mostly deal with blobs that don't go into the centralized release.)

Bottom line, does the separation in https://dev.azure.com/dnceng/internal/_build/results?buildId=306146&view=artifacts raise additional red flags? I'd like to finish this in the next iteration.

dougbu commented:
For example, all of the *_Installers packages have files with the same name as https://dev.azure.com/dnceng/_apis/resources/Containers/2153204/Linux_arm64_Installers?itemPath=Linux_arm64_Installers%2FRelease%2Faspnetcore_base_runtime.version The one that "wins" probably doesn't matter but artifact publication may hit errors. (If we need to merge those artifacts into BlobArtifacts too.)

dougbu commented:
Separately we have a Microsoft.Web.Xdt.Extensions.pdb file in its own Windows_Symbols artifact. Does that need to be moved somewhere to ensure it's uploaded correctly?

The particular file isn't part of a *.symbols.nupkg because the assembly is used exclusively in the AzDO and they get everything through a MSI in our Windows_Installers artifact.

dougbu commented:
(@JunTaoLuo on a mostly-unrelated note: Why don't we have a aspnetcore-targeting-pack for macOS?)

dougbu commented:
More for @riarenas @JohnTortugo @mmitche: There may be a simpler fix here that I'm missing. What exactly are we doing differently in the Extensions repo?

See e.g. https://dev.azure.com/dnceng/internal/_build/results?buildId=305842&view=artifacts That list includes BlobArtifacts and PackageArtifacts but I don't see where those get published, just where they're consumed. (Our pipeline only explicitly creates artifacts named "artifacts", NpmArtifacts, PdbArtifacts and a few logging odds and ends.)

dougbu commented:
Might be getting closer to a resolution here. Found our other repos are passing some or all of the following when in official (aka internal) builds on Windows:
```
          # DotNet-Blob-Feed provides: dotnetfeed-storage-access-key-1
          # Publish-Build-Assets provides: MaestroAccessToken, BotAccount-dotnet-maestro-bot-PAT
          SignType=$(_SignType)
                   /p:TeamName=$(_TeamN- group: DotNet-Blob-Feed
          - group: Publish-Build-Assets
          - name: _OfficialBuildArgs
            value: /p:DotNetSignType=$(_SignType)
                   /p:TeamName=$(_TeamName)
                   /p:DotNetPublishBlobFeedKey=$(dotnetfeed-storage-access-key-1)
                   /p:DotNetPublishBlobFeedUrl=https://dotnetfeed.blob.core.windows.net/aspnet-extensions/index.json
                   /p:DotNetPublishToBlobFeed=$(_DotNetPublishToBlobFeed)
                   /p:DotNetPublishUsingPipelines=$(_PublishUsingPipelines)
                   /p:DotNetArtifactsCategory=$(_DotNetArtifactsCategory)
```

1. This isn't in this PR. To save me the extra builds, is the above enough to get things going?
2. Are the same properties needed when building our Linux and macOS runtime packages and installers?
3. Can any of the above be left out e.g. isn't `$(DotNetPublishBlobFeedUrl)` derived from `$(DotNetArtifactsCategory)` (props in the MSBuild sense here 😸)?

----
Ignore a bit of the above cut 'n paste. We're currently passing `/pSignType=$(_SignType) /p:TeamName=$(_TeamName)` into official builds.

dougbu commented:
Threw everything at it in latest commit. New internal build: https://dev.azure.com/dnceng/internal/_build/results?buildId=306939&view=results

dougbu commented:
@JohnTortugo @riarenas could you please make recommendations based on internal build https://dev.azure.com/dnceng/internal/_build/results?buildId=307941 ? I _think_ I'm setting the publish-related properties correctly but
- Some Linux x64 build failed with a `##[error]Bash exited with code '127'.` (command not found?) message and not much more information (but _some_ Linux builds worked)
- Other Linux builds (perhaps those using a container?) are failing in Publish.proj due to missing directories
- Code sign jobs are failing to publish, likely because the signed .nupkg files aren't placed in the expected location. Can that expected location be overridden? If not, I'll change what our signing project does

riarenas commented:
@johntortugo the errors trying to publish to azure devops build artifacts seems similar than what CoreCLR was facing at some point. Do you remember what the solution was in that case?

dougbu commented:
Made a number of changes and rebased. Latest internal build is: https://dev.azure.com/dnceng/internal/_build/results?buildId=308884

dougbu commented on Directory.Build.props:
```diff
@@ -125,6 +125,11 @@
+  <!-- Include our "loose" PDBs when publishing. -->
+  <ItemGroup>
+    <FilesToPublishToSymbolServer Include="$(SymbolsOutputPath)**\*.pdb" />
+  </ItemGroup>
```
This could also be placed in Publishing.props now. Let's do that if we need another iteration.

dougbu commented:
Latest internal build https://dev.azure.com/dnceng/internal/_build/results?buildId=308894

JunTaoLuo commented on .azure/pipelines/ci.yml:
```diff
@@ -22,515 +22,579 @@ variables:
+  - name: _SignType
+    value: ''
+- ${{ if eq(variables['System.TeamProject'], 'internal') }}:
+  - ${{ if ne(variables['Build.Reason'], 'PullRequest') }}:
```
When do we make PRs to internal? Security fixes?

dougbu replied:
Yup

JunTaoLuo commented on .azure/pipelines/jobs/default-build.yml:
```diff
@@ -75,182 +65,160 @@ parameters:
+      enablePublishUsingPipelines: ${{ variables._PublishUsingPipelines }}
+    enablePublishTestResults: true # publish test results to AzDO (populates AzDO Tests tab)
+    enableTelemetry: true
+    helixRepo: aspnet/AspNetCore-Tooling
```
?

dougbu replied:
yeah, this needs to be fixed.

JunTaoLuo reviewed (approved):
One question otherwise look good.

dougbu commented:
Latest internal build after fixups: https://dev.azure.com/dnceng/internal/_build/results?buildId=308943

dougbu commented:
Last build tonight for me: https://dev.azure.com/dnceng/internal/_build/results?buildId=308951

dougbu commented:
https://dev.azure.com/dnceng/internal/_build/results?buildId=308951 hit almost the same errors as I described [above](https://github.com/aspnet/AspNetCore/pull/13040#issuecomment-521003041). Bit more information:
- Linux x64 build command may have a typo. Alternatively, we might not be running MSBuild in that command.
- Linux MUSL builds aren't much improved though the double-slash issue is resolved. Might be as simple as creating the directory it's complaining about.
- Code sign builds gets an incorrect `$(ArtifactsShippingPackagesDir)` value. Probably best to pass `$(Configuration)` in as a global property in CI builds.
  e.g.
  ```
  Copying file from "F:\workspace\_work\1\a\deps\MacOS_x64_Packages\Release\Shipping\Microsoft.AspNetCore.App.Runtime.osx-x64.3.0.0-preview9.19413.11.symbols.nupkg" to "F:\workspace\_work\1\s\artifacts\packages\Debug\Shipping\Microsoft.AspNetCore.App.Runtime.osx-x64.3.0.0-preview9.19413.11.symbols.nupkg".
  ```
- Code sign builds may need to explicitly tie in Publish.proj from XplatPackageSigner.proj. Publication is not happening.

dougbu commented on .azure/pipelines/jobs/codesign-xplat.yml:
```diff
@@ -26,6 +26,7 @@ jobs:
         solution: eng\tools\XplatPackageSigner\XplatPackageSigner.proj
         msbuildArguments: /p:SignType=$(_SignType)
+          /p:Configuration=Release
           -p:AssetManifestFileName=aspnetcore-${{ parameters.inputName }}-signed.xml
```
Also change this to `/p:` because `-p` sometimes confuses our build scripts and MSBuild on Windows.

dougbu replied:
Separately, docs for the publishing feature are found at https://github.com/dotnet/arcade/blob/master/Documentation/CorePackages/Publishing.md and the main entry point is in https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj

A few possibilities since the `$Configuration)` change probably won't fix the fact no publication is occurring, I suggest you try:
1. Debug the binary log to see whether the properties on the command line aren't overridden or lost somewhere in the `msbuild` world. Fix anything you find.
2. Add `-restore` to the command line i.e. the Arcade SDK might not be imported
3. Hack something together which creates the artifacts\ folder and maybe a few subfolders (if they don't exist for some reason)
4. Add `/verbosity:minimal` or even `/verbosity:quiet` to the command line. This seems like a longshot given how small the console log for this job is. But, @JohnTortugo suggested it because this helped in another repo.

JunTaoLuo commented:
Status: https://dnceng.visualstudio.com/internal/_build/results?buildId=310764 might be an improvement that addresses 1. and 3. Looking into 4. right now and will revisit 2. afterwards.

JunTaoLuo commented:
Huh... my changes didn't help. I won't be able to look at this tomorrow but may be able to revisit on Friday.

wtgodbe commented:
In in the MUSL build:

> /code/build/eng/common/CheckSymbols.ps1

Makes me think that the problem isn't us putting files in the wrong place, but rather somebody looking for files in the wrong place (maybe without a path prefix?) since that file is checked into the repo.

wtgodbe commented:
Fixed the 127 in the Linux build (issue was in build.sh, see https://github.com/aspnet/AspNetCore/pull/13040/commits/0df9195fa3b6ae9170a84df194b8a0f206e23823#diff-0b83f9dedf40d7356e5ca147a077acb4R224)

Regular Linux build is now failing in the same manner as the MUSL builds:

> Path not exist: /code/build/artifacts/../AssetsTmpDir/5d7985bd-934a-433a-a79c-d9a216fa2607/Microsoft.AspNetCore.App.Runtime.linux-x64.3.0.0-preview9.19415.5.nupkg

https://dnceng.visualstudio.com/internal/_build/results?buildId=311919

That path prefix does look correct, as per other paths that get successfully resolved in the binlog, so maybe it's some docker weirdness.

wtgodbe commented:
The `Configuration` fix appears to have worked for that specific issue, but the failure persists.

Before:

> Copying file from "F:\workspace\_work\1\a\deps\MacOS_x64_Packages\Release\Shipping\Microsoft.AspNetCore.App.Runtime.osx-x64.3.0.0-preview9.19414.7.symbols.nupkg" to "F:\workspace\_work\1\s\artifacts\packages\**Debug**\Shipping\Microsoft.AspNetCore.App.Runtime.osx-x64.3.0.0-preview9.19414.7.symbols.nupkg".

After:

>   Copying file from "F:\workspace\_work\1\a\deps\MacOS_x64_Packages\Release\Shipping\Microsoft.AspNetCore.App.Runtime.osx-x64.3.0.0-preview9.19415.5.nupkg" to "F:\workspace\_work\1\s\artifacts\packages\**Release**\Shipping\Microsoft.AspNetCore.App.Runtime.osx-x64.3.0.0-preview9.19415.5.nupkg".

Both failed with:

> Not found SourceFolder: F:\workspace\_work\1\s\artifacts\log\Release\AssetManifest

wtgodbe commented:
I think we're down to 2 issues:

1. Builds in Docker fail during `PublishToAzureDevOpsArtifacts` because they can't find the assets to copy, despite the binlog claiming to have placed the assets there earlier in the build

2. Code sign builds failing because they can't find the Asset Manifest, which should be created during `PublishToAzureDevOpsArtifacts`:https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj#L139-L164. Looks like we download the packages from the corresponding OS's build & sign them, but don't produce another asset manifest.

wtgodbe commented:
@dougbu I am starting to buy into your theory that the Arcade SDK is not being imported. Going to try that angle

wtgodbe commented:
I'm pretty sure that https://github.com/aspnet/AspNetCore/blob/0df9195fa3b6ae9170a84df194b8a0f206e23823/.azure/pipelines/jobs/codesign-xplat.yml#L31 isn't having any effect since we don't restore the Arcade SDK, so Arcade's `publish.proj` doesn't get built. We need to restore the SDK, set the appropriate properties to get the signed packages published (in this case Arcade is going to look for packages in `ArtifactsShippingPackagesDir`: https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj#L61-L63), and invoke publish.proj. @riarenas @dougbu sound right? I'm working on wiring that up

dougbu commented on .azure/pipelines/jobs/codesign-xplat.yml:
```diff
@@ -21,11 +21,21 @@ jobs:
         downloadPath: $(Build.StagingDirectory)/deps/
         itemPattern: '**/*.nupkg'
+    - script: powershell -ExecutionPolicy ByPass -NoProfile build.ps1 /t:Restore
+      displayName: Restore tools
```
1. checked locally and .\restore.cmd (the equivalent of this command) does not restore anything for the XplatPackageSigner.proj project
    - suggest adding `-restore` to the `msbuildArguments` below
2. The XplatPackageSigner.proj project does not have a 'restore' target
    - might need to rename the file to XplatPackageSigner.**cs**proj and maybe make a few other changes e.g. add `<IsPackable>false</IsPackable>`, `<IsProjectReferenceProvider>false</IsProjectReferenceProvider>` and `<IsShipping>false</IsShipping>`
    - alternatively, might be able to `<import/>` the correct NuGet files to add a 'restore' target

dougbu commented on .azure/pipelines/jobs/codesign-xplat.yml:
```diff
@@ -21,11 +21,21 @@ jobs:
       path: $(Build.StagingDirectory)\deps\${{ parameters.inputName }}_Packages\
+    - name: ${{ parameters.inputName }}_Logs
+      path: artifacts/log/
+      publishOnError: true
```
For good hygiene, always put the log upload first

wtgodbe reviewed (changes requested):
https://github.com/aspnet/AspNetCore/pull/13040/commits/3247a9c46284aa341dd3c3c23776f25b4b6dff3f is a hack to work around dot.net being down, don't merge this until that's reverted

wtgodbe commented:
Latest build: https://dnceng.visualstudio.com/internal/_build/results?buildId=313424

JunTaoLuo commented:
/azp run all

wtgodbe commented:
https://dnceng.visualstudio.com/internal/_build/results?buildId=313848 latest build is still failing in the same way when trying to restore the SDK:

> ##[error]A compatible SDK version for global.json version: [3.0.100-preview7-012821] from [F:\workspace\_work\1\s\global.json] was not found
##[error]eng\tools\XplatPackageSigner\XplatPackageSigner.csproj(0,0): Error : Unable to locate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.

On Linux, restore.cmd itself is failing:

> C:\Users\vsagent\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19411.1\tools\Tools.proj : error : Unable to locate the .NET Core SDK. Check that it is installed and that the version specified in global.json (if any) matches the installed version.
C:\Users\vsagent\.nuget\packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19411.1\tools\Tools.proj : error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.

dougbu commented on .azure/pipelines/jobs/codesign-xplat.yml:
```diff
@@ -14,18 +14,34 @@ jobs:
         artifactName: ${{ parameters.inputName }}_Packages
         downloadPath: $(Build.StagingDirectory)/deps/
         itemPattern: '**/*.nupkg'
+    - script: echo "##vso[task.prependpath]$BUILD_SOURCESDIRECTORY/.dotnet"
```
Minor but you might as well use `$(DOTNET_HOME)` here.

wtgodbe reviewed (changes requested):
https://github.com/aspnet/AspNetCore/pull/13040/commits/0da1d0a6b7083a1d8b69eab37208e8c4085cee5b includes a change to upload packages in dockerbuilds even on error, as an experiment to test a theory. That change should be removed before this is merged

wtgodbe commented:
https://dnceng.visualstudio.com/internal/_build/results?buildId=317084 new build

JunTaoLuo commented:
Latest internal build: https://dnceng.visualstudio.com/internal/_build/results?buildId=322948

wtgodbe reviewed (approved):
All workarounds have been removed

wtgodbe commented:
Code Sign job is now working! Unfortunately the docker builds are still failing to upload the packages to PackageArtifacts (and their Asset Manifests). This is the error we thought was fixed by setting verbosity:quiet. @riarenas and @JohnTortugo have informed me that they are working on a fix for this on the Arcade side, but until then we are probably stuck (CC @dougbu @JunTaoLuo)

Although, it occurs to me that since the CodeSign jobs are actually re-uploading the same packages that the dockerbuild jobs failed to upload, we might be able to get away with taking this in its current hacky state, until we get a real fix from Arcade. This works since the CodeSign jobs download the Unix packages from Artifact locations that we upload to manually, then re-upload the packages to PackageArtifacts, where the dockerbuild jobs failed to publish to. Might be best to just wait for the real fix though, since the current state is pretty ugly/confusing 😆

wtgodbe commented:
Issue for fixing the dockerfiles: https://github.com/aspnet/AspNetCore-Internal/issues/3043

dougbu commented:
Latest internal build _built_ successfully but failed signing and `sourcelink` validation. The first looks like the publication of signed packages for non-Windows builds silently failed. The second is **Just Weird:tm:** and bears investigation.

@riarenas @JohnTortugo are the PowerShell erorrs in https://dnceng.visualstudio.com/internal/_build/results?buildId=329231&view=logs&s=4304c483-acc4-5f4c-d091-add65b819655&j=0a6c679f-72be-5867-422e-acb741a068a3 known problems?

dougbu commented:
I force-pushed this branch to pick up the latest from 'release/3.0'. Don't worry, I fetched first 😸

dougbu commented:
Sorry, my force-push broke something important I think. Previous commit at tip in internal repo was d7dab797de I'm signing off for the night. But, was GitHub missing commits you'd added internally @wtgodbe?

dougbu commented:
Redid the rebase and it didn't change anything. Looking&hellip;

dougbu commented:
New internal build: https://dev.azure.com/dnceng/internal/_build/results?buildId=330528

wtgodbe commented:
I think what might be happening is that, since we're manually calling into the Microbuild signing task ourselves, Microbuild is just signing the .nupkg & not the .dll's within? If that's the case, we could download the nupkg's to `ArtifactsShippingPackagesDir` & call `.\eng\common\build.ps1 - sign` before we publish. Arcade's signing target cracks open the .nupkg's & signs the files within.

JohnTortugo commented:
AFAIU the microbuild plugin doesn't extract and process containers. You have to do that yourself. The Arcade SignToolTask does that before submitting the files to MicroBuild.

wtgodbe commented:
All unsigned falls are Unix runtime files, so seems like my theory from above was true. Will make the change & start a new build.

wtgodbe commented:
https://github.com/aspnet/AspNetCore/pull/13040/commits/02e096819a280e6beeeadbc050bd2a1190b37101 uses Arcade signing, and deletes XPlatPackageSigner as it won't be needed anymore if the change works. New build: https://dnceng.visualstudio.com/internal/_build/results?buildId=330774

wtgodbe commented:
> error : Could not determine certificate name for signable file(s):
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\Sign.proj(81,5): error : 	File: F:\workspace\_work\1\s\artifacts\tmp\Release\ContainerSigning\4\runtimes/linux-arm64/lib/netcoreapp3.0/Microsoft.Extensions.Caching.Abstractions.dll 

Looking

wtgodbe commented:
Build where I manually set the Cert for .dll's to `Microsoft400`: https://dnceng.visualstudio.com/internal/_build/results?buildId=330992

Build where I do the above, and attempt to fix the dockerfiles to resolve https://github.com/dotnet/core-eng/issues/7657: https://dnceng.visualstudio.com/internal/_build/results?buildId=331044

@riarenas @JohnTortugo PTAL at https://github.com/aspnet/AspNetCore/pull/13040/commits/9a7d3ce998606ef3b3ba5418a0205cb9c63b74c9

wtgodbe commented:
Looks like the signing fix worked, but the docker fix didn't

wtgodbe reviewed (changes requested):
If we fix the docker files in this PR, we should turn the verbosity back up for dockerbuilds as well.

wtgodbe commented:
New build with Verbosity back to normal: https://dnceng.visualstudio.com/internal/_build/results?buildId=331111

wtgodbe reviewed (approved):
Verbosity is back to normal, and the docker builds are successfully publishing artifacts.

JunTaoLuo commented on global.json:
```diff
@@ -1,6 +1,7 @@
   "sdk": {
-    "version": "3.0.100-preview8-013656"
+    "version": "3.0.100-preview8-013656",
+    "rollforward": "major"
```
So are we removing this?

wtgodbe commented:
> So are we removing this?

Yeah, if the current build (https://dnceng.visualstudio.com/internal/_build/results?buildId=331132) is green, let's do it in a cleanup PR

wtgodbe commented:
We should try to find out if there's a way to ensure we don't double-sign things that we shouldn't be (e.g. 3rd party libraries)

wtgodbe commented:
Newest build where I fixed the CodeSign jobs doing DryRun signing - https://dnceng.visualstudio.com/internal/_build/results?buildId=331434

@dougbu if that build is green, this should be ready for merge

dougbu commented:
@wtgodbe that build failed signing validation because the macOS and Linux assemblies still weren't signed. It looks like one of the following
- the signing jobs aren't cracking open the .nupkg files to sign assemblies (still)
- the signed assemblies are getting published to a different location from the unsigned ones and verification chooses to use the earlier manifest e.g. a macOS manifest rather than one from the code signing job for macOS assets
- the manifests generated in our xplat signing jobs are corrupt and being ignored

wtgodbe commented:
The issue is that the code-sign jobs aren't uploading their packages as PackageArtifacts, but as BlobArtifacts. Will try to figure out why.

wtgodbe commented:
It's because we download the packages to:

> F:\workspace.1\_work\1\s\artifacts\packages\Release\shipping\MacOS_x64_Packages\Release\Shipping\Microsoft.AspNetCore.App.Runtime.osx-x64.3.0.0-rc1.19429.12.nupkg

While Arcade looks for packages directly in the root of the package dir: https://github.com/dotnet/arcade/blob/682631e1d708c9de2dacef8b836b9f73fcdd3c87/src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj#L61

Will look into changing the download path

wtgodbe commented:
New build: https://dnceng.visualstudio.com/internal/_build/results?buildId=332420

wtgodbe commented:
I've had to retry the Templates tests twice now w/ unrelated failures, so validation is taking forever 😢

wtgodbe commented:
Internal job where I disabled tests, just to speed things along: https://dnceng.visualstudio.com/internal/_build/results?buildId=332925

wtgodbe commented:
4 failures in signing validation: apphost.exe (4 instances), 86 `.js` files, and the following 4:

> [File] WixDepCA, Signed: False [Error] HRESULT: 800b0100 (No signature was present in the subject)
> [File] WixUIWixca, Signed: False [Error] HRESULT: 800b0100 (No signature was present in the subject)
> [File] WixDepCA, Signed: False [Error] HRESULT: 800b0100 (No signature was present in the subject)
> [File] WixUIWixca, Signed: False [Error] HRESULT: 800b0100 (No signature was present in the subject)

The first 2 are due to https://github.com/aspnet/AspNetCore/blob/5279b7d6adc6d90567212d15108d34c455163414/eng/Signing.props#L78-L79 & https://github.com/aspnet/AspNetCore/blob/5279b7d6adc6d90567212d15108d34c455163414/eng/Signing.props#L85-L86. It's easy enough to remove those exclusions, though I don't know if we'd rather somehow exclude them from signcheck (@riarenas @JohnTortugo please advise on if this is possible)

As for the Wix files, I'm not sure what we want to do with those. Probably opt into signing them? I'll push a PR that includes all 3 of these things in signing, and leave it to @JunTaoLuo & @dougbu to decide if that's the fix we want.

wtgodbe commented:
I _believe_ the changes to `Signing.props` in https://github.com/aspnet/AspNetCore/pull/13040/commits/8ea405ff77387aabba172678300f996c50118c15 should exclude the files I mentioned in https://github.com/aspnet/AspNetCore/pull/13040#issuecomment-526780052 from SignCheck. If so, then @JunTaoLuo and @dougbu please decide if `apphost.exe`, `*.js`, and `Wix*` files should be signed, or excluded from signcheck, and make the corresponding change to Signing.props. I'm optimistic that this is the last change needed... 

New build: https://dnceng.visualstudio.com/internal/_build/results?buildId=333150

dougbu committed:
Cleanup hopefully-unnecessary YAML changes
- remove `$(DOTNET_HOME)` settings
- do not use `activate.ps1` or `restore.cmd` in codesign-xplat.yml

nit:
- restore deleted blank line
- move log artifact first in codesign-xplat.yml too

dougbu committed:
Correct syntax of `@(SignValidationExclusionList)` items

dougbu commented:
Squashed all of the commits and rebased. Then, added a bit of cleanup and corrected the sign-check exclusions.

dougbu commented:
Test build: https://dev.azure.com/dnceng/internal/_build/results?buildId=334057

dougbu commented:
Internal test build went pretty well except
1. I added a default channel for this branch and '.NET Tools - Validation'. This lead to publish-build-assets.yml failing
    ```
    D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5): error : RestApiException`1: The response contained an invalid status code 500 Internal Server Error
    D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5): error : 
    D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5): error : Body: {"message":"An error occured.","errors":null}
    D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5): error :    at Microsoft.DotNet.Maestro.Client.Builds.OnCreateFailed(HttpRequestMessage req, HttpResponseMessage res) in /_/src/Maestro/Client/src/Generated/Builds.cs:line 234
    D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5): error :    at Microsoft.DotNet.Maestro.Client.Builds.CreateInternalAsync(BuildData body, CancellationToken cancellationToken) in /_/src/Maestro/Client/src/Generated/Builds.cs:line 293
    D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5): error :    at Microsoft.DotNet.Maestro.Client.Builds.CreateAsync(BuildData body, CancellationToken cancellationToken) in /_/src/Maestro/Client/src/Generated/Builds.cs:line 214
    D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5): error :    at Microsoft.DotNet.Maestro.Tasks.PushMetadataToBuildAssetRegistry.PushMetadataAsync(CancellationToken cancellationToken) in /_/src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:line 87
    D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5): error : 
    ##[error].packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5): error : RestApiException`1: The response contained an invalid status code 500 Internal Server Error

    Body: {"message":"An error occured.","errors":null}
      at Microsoft.DotNet.Maestro.Client.Builds.OnCreateFailed(HttpRequestMessage req, HttpResponseMessage res) in /_/src/Maestro/Client/src/Generated/Builds.cs:line 234
      at Microsoft.DotNet.Maestro.Client.Builds.CreateInternalAsync(BuildData body, CancellationToken cancellationToken) in /_/src/Maestro/Client/src/Generated/Builds.cs:line 293
      at Microsoft.DotNet.Maestro.Client.Builds.CreateAsync(BuildData body, CancellationToken cancellationToken) in /_/src/Maestro/Client/src/Generated/Builds.cs:line 214
      at Microsoft.DotNet.Maestro.Tasks.PushMetadataToBuildAssetRegistry.PushMetadataAsync(CancellationToken cancellationToken) in /_/src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:line 87

    Build FAILED.
    ```
    @JohnTortugo @mmitche @riarenas does this mean the pipeline or the relevant build step lacks a variable group?
2. Need to add conditions to the publish-build-assets.yml step so that it doesn't run when test jobs fail. For example, the build attempted to publish to the BAR though the "Test: Windows Server 2016 x64" job failed.

dougbu commented:
In addition

3. The internal and public builds failed the same `ServerComparison.FunctionalTests.HelloWorldTests.HelloWorld` test variants. Need to help that test group find `dotnet`. On the CI, these tests currently look for 'C:\\Users\\vsagent\\.dotnet\\x86\\dotnet.exe'
    Failing variants:
    - HelloWorld-Server: HttpSys, TFM: netcoreapp3.0, Type: Portable, Arch: x86
    - HelloWorld-Server: IISExpress, TFM: netcoreapp3.0, Type: Portable, Arch: x86, Host: InProcess
    - HelloWorld-Server: IISExpress, TFM: netcoreapp3.0, Type: Portable, Arch: x86, Host: OutOfProcess
    - HelloWorld-Server: Kestrel, TFM: netcoreapp3.0, Type: Portable, Arch: x86

dougbu committed:
Use `$env:DOTNET_INSTALL_DIR` in `DotNetCommands`
- relax expectations that an arch-specific folder exists under (say) `$env:DOTNET_HOME`
- avoids need to define `$env:DOTNET_HOME` in all bjos on CI

dougbu committed:
Do not publish to the BAR unless tests succeed

dougbu commented:
Another new build: https://dev.azure.com/dnceng/internal/_build/results?buildId=334541

dougbu committed:
!fixup! Move BAR publication to the end correctly
- move that job to the end of the Build stage in the YAML too

dougbu commented:
Last build had a syntax error and didn't hold BAR publication to the end. New internal build: https://dev.azure.com/dnceng/internal/_build/results?buildId=334866

dougbu commented:
In we're down to two issues:
1. Windows and non-Windows don't seem to be included in BlobArtifacts or PackageArtifacts
2. Publish to the BAR is failing still:
    ```
    .packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5):   error : RestApiException`1: The response contained an invalid status code 500
    Internal Server Error Body: {"message":"An error   occured.","errors":null} at
    Microsoft.DotNet.Maestro.Client.Builds.OnCreateFailed(HttpRequestMessage   req, HttpResponseMessage res) in   /_/src/Maestro/Client/src/Generated/Builds.cs:line 234 at
    Microsoft.DotNet.Maestro.Client.Builds.CreateInternalAsync(BuildData body,   CancellationToken cancellationToken) in /_/src/Maestro/Client/src/Generated/Builds.cs:line   293 at
    Microsoft.DotNet.Maestro.Client.Builds.CreateAsync(BuildData body,   CancellationToken cancellationToken) in   /_/src/Maestro/Client/src/Generated/Builds.cs:line 214 at
 
 
 
 Microsoft.DotNet.Maestro.Tasks.PushMetadataToBuildAssetRegistry.PushMetadataAsync(CancellationToken   cancellationToken) in /_/src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:line   87
    --
    Build failed.
    ```

dougbu commented:
For artifacts with content not listed in the created manifests or the Arcade BlobArtifacts, see https://dev.azure.com/dnceng/internal/_build/results?buildId=334866&view=artifacts:
- Linux_arm_Installers
- Linux_arm64_Installers
- Linux_musl_arm64_Installers
- Linux_musl_x64_Installers
- Linux_x64_Installers
- MacOS_x64_Installers
- Windows_arm_Installers
- Windows_Installers

Most installers are named aspnetcore-runtime-{version}.tar.gz or aspnetcore-runtime-internal-{version}.tar.gz but other extensions e.g. .deb, .exe, .msi and .zip are also used.

dougbu commented on eng/Publishing.props:
```diff
@@ -0,0 +1,11 @@
+
+  <!-- Include our "loose" PDBs when publishing. -->
+  <ItemGroup>
+    <FilesToPublishToSymbolServer Include="$(SymbolsOutputPath)**\*.pdb" />
```
This line did not add the expected Microsoft.Web.Xdt.Extensions.pdb file to PDBArtifacts. Nothing logged about uploading that file in https://dev.azure.com/dnceng/internal/_build/results?buildId=334866&view=logs&s=6884a131-87da-5381-61f3-d7acc3b91d76&j=1b89928a-2219-5ef9-602f-f95beb3da4dc and the build produced no PDBArtifacts.

@riarenas @JohnTortugo @mmitche should we be setting `$(UsingToolSymbolUploader)` as well or instead of adding this item? Or, is placing the file in the PDBArtifacts directory our only option?

dougbu commented:
@JohnTortugo @riarenas @mmitche I probably need to change the following properties:
``` xml
<InstallersOutputPath>$(ArtifactsDir)installers\$(Configuration)\</InstallersOutputPath>
<SymbolsOutputPath>$(ArtifactsDir)symbols\$(Configuration)\</SymbolsOutputPath>
```
(`installers/` and `symbols/` are sibling directories to `packages/`.)

Should that snippet instead be the following?
``` xml
<InstallersOutputPath>$(ArtifactsDir)BlobArtifacts\$(Configuration)\</InstallersOutputPath>
<SymbolsOutputPath>$(ArtifactsDir)PDBArtifacts\$(Configuration)\</SymbolsOutputPath>
```

Is there a Shipping / NonShipping or other convention within either `artifacts/` directory?

----
The above won't fix the BAR publication failures. Any suggestions there much appreciated❕

mmitche commented:
> @JohnTortugo @riarenas @mmitche I probably need to change the following properties:
> 
> ```
> <InstallersOutputPath>$(ArtifactsDir)installers\$(Configuration)\</InstallersOutputPath>
> <SymbolsOutputPath>$(ArtifactsDir)symbols\$(Configuration)\</SymbolsOutputPath>
> ```
> 
> (`installers/` and `symbols/` are sibling directories to `packages/`.)
> 
> Should that snippet instead be the following?
> 
> ```
> <InstallersOutputPath>$(ArtifactsDir)BlobArtifacts\$(Configuration)\</InstallersOutputPath>
> <SymbolsOutputPath>$(ArtifactsDir)PDBArtifacts\$(Configuration)\</SymbolsOutputPath>
> ```
> 
> Is there a Shipping / NonShipping or other convention within either `artifacts/` directory?

I'm not entirely sure.  Sounds right? @JohnTortugo will know better.

JohnTortugo commented:
Talked with @dougbu offline.

riarenas commented:
> Publish to the BAR is failing still:

I was able to repro this error by having a default channel set up for the internal and github version of the repo. I filed https://github.com/dotnet/arcade/issues/3826.
You should be able to unblock this by deleting the default channel mapping for the internal repo while we fix that.

dougbu commented:
> You should be able to unblock this by deleting the default channel mapping for the internal repo while we fix that.

Does this mean it won't be necessary to remove duplication between what we publish in e.g. the 'Build: Linux arm' and 'Code-sign Linux_arm packages' jobs? That simplifies things if so.

riarenas commented:
If there is any other inconsistency in the manifests the step might still fail, so I'm not sure what you mean. This particular error:

```
.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19425.1\tools\SdkTasks\PublishBuildAssets.proj(30,5):   error : RestApiException`1: The response contained an invalid status code 500
Internal Server Error Body: {"message":"An error   occured.","errors":null} at
Microsoft.DotNet.Maestro.Client.Builds.OnCreateFailed(HttpRequestMessage   req, HttpResponseMessage res) in   /_/src/Maestro/Client/src/Generated/Builds.cs:line 234 at
Microsoft.DotNet.Maestro.Client.Builds.CreateInternalAsync(BuildData body,   CancellationToken cancellationToken) in /_/src/Maestro/Client/src/Generated/Builds.cs:line   293 at
Microsoft.DotNet.Maestro.Client.Builds.CreateAsync(BuildData body,   CancellationToken cancellationToken) in   /_/src/Maestro/Client/src/Generated/Builds.cs:line 214 at
Microsoft.DotNet.Maestro.Tasks.PushMetadataToBuildAssetRegistry.PushMetadataAsync(CancellationToken   cancellationToken) in /_/src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:line   87

Build failed.
```
Looks to be related to the multiple default channels.

JohnTortugo commented:
@dougbu I suggest that if possible you try it first without removing the asset duplication thing that we discussed. The duplication of default channel is frequently a source of this kind of problem and it might as well been for your build to.

dougbu committed:
Correct publication of loose .pdb files and use Arcade to publish installers

dougbu commented:
@JohnTortugo @riarenas please have a look at commit 6ad335f90cf8. I greatly appreciate your help simplifying this today❕

----
Latest internal build is https://dev.azure.com/dnceng/internal/_build/results?buildId=337728.

dougbu commented on eng/Publishing.props:
```diff
@@ -0,0 +1,51 @@
+
+  <Target Name="_PublishInstallers">
+    <!-- This target is defined in eng/targets/Packaging.targets and included in every C# and F# project. -->
+    <MSBuild Projects="$(RepoRoot)src\Mvc\Mvc\src\Microsoft.AspNetCore.Mvc.csproj"
```
Choosing this project was fairly random😈

dougbu commented on src/Hosting/Server.IntegrationTesting/src/Common/DotNetCommands.cs:
```diff
@@ -16,8 +16,13 @@ public static class DotNetCommands
         // Compare to https://github.com/aspnet/BuildTools/blob/314c98e4533217a841ff9767bb38e144eb6c93e4/tools/KoreBuild.Console/Commands/CommandContext.cs#L76
         public static string GetDotNetHome()
         {
+            // runtest.* scripts throughout the repo define $env:DOTNET_HOME
```
@jkotalik the changes to this file are an attempt to avoid defining `$env:DOTNET_HOME` globally (see https://github.com/aspnet/AspNetCore/pull/13040/files#diff-f3d39efe7528a28e4995c18938c8a3d3L117) when it has _no_ impact outside 3 or 4 of our tests. But, I'll revert this if this file is actually shipped. (No need for unnecessary product changes at this point in 3.0.)

How is this file used outside our own tests?

jkotalik replied:
It shouldn't be used outside of our tests. It's in the integration testing project.

dougbu committed:
!fixup! Qualify a metadata reference, remove a quote !fixup!

dougbu commented:
Another internal build: https://dev.azure.com/dnceng/internal/_build/results?buildId=338121

dougbu commented:
@mmitche @riarenas @JohnTortugo, the `PushToAzureDevOpsArtifacts` task is [now failing](https://dev.azure.com/dnceng/internal/_build/results?buildId=338121) with errors like
```
The file is not a valid nupkg. File path: /Users/vsts/agent/2.155.1/work/1/s/artifacts/installers/Release/aspnetcore-runtime-3.0.0-rc1.19454.10-osx-x64.tar.gz
```
```
Nuspec file does not exist in package.
```

I believe the issue is the Arcade SDK's Publish.proj hard-codes `PublishFlatContainer="false"` and we're now publishing items other than .nupkg files. The only workarounds I see are to add `<ExcludeFromManifest>true</ExcludeFromManifest>` metadata or to fall back to the [arcade/toolset approach](https://github.com/dotnet/toolset/blob/a353eb70a5a06659c328f1c90d1c85f8b28eec21/eng/Publishing.props#L41-L50) of calling the `PushToAzureDevOpsArtifacts` task outside Publish.proj. (A simple change to use `PublishFlatContainer="false"` in Publish.proj would disable creation of PackageArtifacts and that seems bad.) Is there another approach?

I'm going to try `<ExcludeFromManifest>true</ExcludeFromManifest>` because our manifests didn't previously list the installers.

dougbu commented:
Latest trial: https://dev.azure.com/dnceng/internal/_build/results?buildId=338164

dougbu commented:
Weird, Source Build job is now failing with the errors we saw earlier today @JunTaoLuo. Did something change to turn those ignored errors into build failures?

dougbu commented:
`%(ExcludeFromManifest)` avoided some errors but _prevented_ publication of the installers in BlobArtifacts. (No longer clear to me what the metadata is intended for.) Is a separate use of the `PushToAzureDevOpsArtifacts` task the only option for publishing our installers?

dougbu commented:
@riarenas @JohnTortugo @mmitche see dotnet/arcade#3855 for one way to unblock this without duplicating some of Publish.proj in our Publishing.props.

dougbu committed:
Use new `%(ItemsToPushToBlobFeed.PublishFlatContainer)` metadata

dougbu commented:
:up::date: to
- rebase branch and resolve conflicts w/ 'release/3.0' changes
  - pick up .NET Core SDK pinning to unbreak source build jobs
- update Arcade dependencies to dotnet/arcade@dd593acc8b08 fix
- use new Arcade metadata to unbreak installer publication

dougbu committed:
!fixup! bad merge of ci.yml and `$(ManifestArtifactData)` value

dougbu commented:
New internal build: https://dev.azure.com/dnceng/internal/_build/results?buildId=340770

@Pilchie if that succeeds, should I rebase this PR on 'release/3.0-preview9' tomorrow morning?

Pilchie commented:
@dougbu I don’t think we need this in the P9 branch for servicing readiness experiments, but double check with @mwilkie.

Pilchie commented:
(Got that handle wrong, sorry I’m on mobile)

mmitche commented:
> New internal build: https://dev.azure.com/dnceng/internal/_build/results?buildId=340770
> 
> @Pilchie if that succeeds, should I rebase this PR on 'release/3.0-preview9' tomorrow morning?

Nope. Stay in 3.0

dougbu commented:
> Stay in 3.0

Internal build failed anyhow. Doesn't look like my Arcade fix (dotnet/arcade@dd593acc8b08) is available even though the Arcade build used in this PR (dotnet/arcade@e99f81b0e328) is _later_. @mmitche what else needs to happen besides the merge, official Arcade build, and the dependency update in this PR? For example, does this branch need an Arcade build that includes its looped dependency update (dotnet/arcade@3c963ed6a079)?

mmitche commented:
No, the default versions of the feed tasks package will need to be updated in arcade, or your build can directly add a reference to the feed tasks package which override the default version, like this has:

https://github.com/dotnet/core-setup/blob/master/eng/Version.Details.xml#L130

dougbu commented:
Not particularly useful yet. Waiting for a successful AspNetCore-Tooling build that is branded correctly.

dougbu commented:
@mmitche my latest internal build (https://dev.azure.com/dnceng/internal/_build/results?buildId=341787) included such an entry in the Version.Details.xml file as well as a property in Versions.props. But the build failed in exactly the same fashion.

My suspicion is the Directory.Build.props file that's in the same directory as Publish.proj in the Arcade SDK prevents anything seeing the updated package version. I named the property incorrectly for Arcade (it's `$(MicrosoftDotNetBuildTasksFeedVersion)` in the Arcade SDK) but it wouldn't have mattered. Without going all custom e.g. https://github.com/dotnet/core-setup/tree/002bc9ef5d0d6d433321df325b5928aaf80282a4/publish, as I'm trying to avoid, what are the options?

Feels like we're blocked until Arcade updates to a later version of the Tasks.Feed package.

dougbu committed:
Update dependencies from dotnet/arcade build '20190908.2'
- upgrade to eg. Arcade SDK '1.0.0-beta.19458.2' package version
- pick up dotnet/arcade@dd593acc8b08 fix
  - enable use of `%(PublishFlatContainer)` metadata
- also corrects issues w/ signing validation

dougbu commented:
:up::date: to remove commits I'd otherwise revert. Use latest Arcade bits.

New internal build: https://dev.azure.com/dnceng/internal/_build/results?buildId=342736

dougbu commented:
@mmitche @riarenas @JohnTortugo I could use some help here. The latest build is https://dev.azure.com/dnceng/internal/_build/results?buildId=344699 and it's failing in the same fashion as previous builds though I just updated to the latest Arcade that's in the channel.

This build used the 2.2.0-beta.19454.31 version of the Microsoft.DotNet.Build.Tasks.Feed package according to the binlog at https://dev.azure.com/dnceng/_apis/resources/Containers/2343395?itemPath=Linux_x64_Logs%2Fbuild.rpm.binlog. But, Maestro++ says the Arcade 20190908.1 build (Arcade SDK v1.0.0-beta.19458.1) depended on Microsoft.DotNet.Build.Tasks.Feed from the 20190907.1 build (v2.2.0-beta.19457.1).

Why is `$(MicrosoftDotNetBuildTasksFeedVersion)` still '2.2.0-beta.19454.31'?

dougbu commented:
Another way to ask the question: Which needs to be manually updated:
- https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/Versions.props
- https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/DefaultVersions.props

(And, why aren't these files auto-generated?)

dougbu commented:
OIC, it's both. I'll model a PR after @riarenas's dotnet/arcade#3852&hellip;

dougbu commented:
Opened dotnet/arcade#3898

We need to wait for that to get in, an Arcade build, and subsequent BAR update before updating dependencies and trying again here ☹️

dougbu commented:
Should watch dotnet/arcade#3897 instead. Thanks @mmitche❕

dougbu commented:
Latest build was https://dev.azure.com/dnceng/internal/_build/results?buildId=344913 It got all the way to Validate Signing Validation and failed with a set of errors we've seen before. Either eng/Signing.props is the wrong place for `@(SignValidationExclusionList)` items or the syntax I'm using is still wrong.

@joeloff @riarenas @JohnTortugo @mmitche what's incorrect about https://github.com/aspnet/AspNetCore/pull/13040/files#diff-3f0380cc81441e712c5eba1d7b77e649R88-R96❔

dougbu commented:
Re-checking eng/Common/SigningValidation.proj, I'm not sure how it could find eng/Signing.props. The file it's looking for is `$(MSBuildThisFileDirectory)Signing.props`. Shouldn't it look for Signing.props in `$(RepositoryEngineeringDir)`❔

JohnTortugo commented:
> Latest build was https://dev.azure.com/dnceng/internal/_build/results?buildId=344913 It got all the way to Validate Signing Validation and failed with a set of errors we've seen before. Either eng/Signing.props is the wrong place for `@(SignValidationExclusionList)` items or the syntax I'm using is still wrong.

The SignValidationExclusion list is actually a .txt file. It's referenced in the post-build scripts here: https://github.com/dotnet/arcade/blob/master/eng/common/templates/post-build/post-build.yml#L64 . [See here how core-setup uses it](https://github.com/dotnet/core-setup/blob/master/eng/SignCheckExclusionsFile.txt).

JohnTortugo commented:
> Re-checking eng/Common/SigningValidation.proj, I'm not sure how it could find eng/Signing.props. The file it's looking for is `$(MSBuildThisFileDirectory)Signing.props`. Shouldn't it look for Signing.props in `$(RepositoryEngineeringDir)`❔

Ignore this file. This isn't used by the post-build stages. This is part of an old infrastructure that will be deprecated once all repos have onboarded in post-build stages and/or left the old infra.

dougbu committed:
Handle signing validation exclusions in SignCheckExclusionsFile.txt file

dougbu committed:
Use `$(DotNetFinalVersionKind)` in preparation for servicing builds
- set `$(IsStableBuild)` for use in Arcade infrastructure

dougbu commented:
@mmitche my last update also addresses your `$(IsStableBuild)` / `$(DotNetFinalVersionKind)` ask.

dougbu commented:
@JohnTortugo thanks for the info. I created the correct file.

dougbu committed:
!fixup! typos in last two commits
- `$(IsStableBuild)` is not always defined
- remove extra angle bracket
- match Arcade versions in Version.Details.xml with those elsewhere

dougbu commented:
@joeloff @riarenas @JohnTortugo switch to use eng/SignCheckExclusionsFile.txt changed nothing. The right parameters were passed to the sign check tool in https://dev.azure.com/dnceng/internal/_build/results?buildId=346960 but validations still failed.

I suspect the [syntax I used](https://github.com/aspnet/AspNetCore/blob/dc965420ef43b5ad9589d2d9afe14fe24cefb8f5/eng/SignCheckExclusionsFile.txt) was wrong but only the two entries like `WixDepCA;*.msi;` look incorrect because the filename portion _requires_ a leading wildcard. I also don't think the tool supports nested parents correctly (the relevant *.msi files are packed within .nupkg files).

How would I get apphost.exe and the *.js files excluded? (I recognize the `**` in the first line is unnecessary and the font files are excluded by default.)

joeloff commented:
let me take a look at the nupkg

joeloff commented:
So looks like this might be a regression from when the new logic for handling parents were introduced. You can explicitly exclude these files for now. I'll get a fix into Arcade for the built-in exclusions

dougbu commented:
> You can explicitly exclude these files for now.

Not sure what you mean. Do we need to exclude the 98 affected files one-by-one? In particular, why isn't the simple `*.js;;` exclusion working?

joeloff commented:
Oh, for those, you need to just update them to *.js;*;

dougbu commented:
So, the following should work?
```
apphost.exe;; Exclude the apphost because this is expected to be code-signed by customers after the SDK modifies it.
.js;; We do not sign JavaScript files.
.binlog;; MSBuild binary logs are not signed though they are sometimes placed where validation thinks they should be.
WixUIWixca|WixDepCA;; We do not sign WiX content in our installers.
```

(I don't see how the apphost.exe exclusion could work given the check passes a fairly-random filename into `Exclusions.IsMatch(...)`.)

dougbu committed:
Update SignCheckExclusionsFile.txt syntax but disable signing validation for now
- validation may have other bugs and we should investigate separately

dougbu committed:
Remove incorrectly-added file

dougbu commented:
Disabled signing validation for now. Filed #13864 to cover re-enabling these checks.

dougbu commented:
Internal build https://dev.azure.com/dnceng/internal/_build/results?buildId=347936 got all the way to the 'Publish Assets' step in '.NET Tools - Validation Publishing' but failed there:
```
MSBuild auto-detection: using msbuild version '16.200.19.32702' from 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\bin'.
  Accepted https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/d1622942-d16f-48e5-bc83-96f4539e7601/nuget/v2/ 6098ms
Your package was pushed.
  Pushing 0 packages.
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19458.2\tools\SdkTasks\PublishArtifactsInManifest.proj(65,5): error : Sequence contains more than one element
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19458.2\tools\SdkTasks\PublishArtifactsInManifest.proj(65,5): error :    at System.Linq.ThrowHelper.ThrowMoreThanOneElementException()
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19458.2\tools\SdkTasks\PublishArtifactsInManifest.proj(65,5): error :    at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source)
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19458.2\tools\SdkTasks\PublishArtifactsInManifest.proj(65,5): error :    at Microsoft.DotNet.Build.Tasks.Feed.PublishArtifactsInManifest.PublishBlobsToAzureStorageNugetFeedAsync(List`1 blobsToPublish, IMaestroApi client, Build buildInformation, FeedConfig feedConfig)
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19458.2\tools\SdkTasks\PublishArtifactsInManifest.proj(65,5): error :    at Microsoft.DotNet.Build.Tasks.Feed.PublishArtifactsInManifest.HandleBlobPublishingAsync(IMaestroApi client, Build buildInformation) in /_/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifest.cs:line 497
D:\a\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19458.2\tools\SdkTasks\PublishArtifactsInManifest.proj(65,5): error :    at Microsoft.DotNet.Build.Tasks.Feed.PublishArtifactsInManifest.ExecuteAsync() in /_/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifest.cs:line 199
##[error].packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19458.2\tools\SdkTasks\PublishArtifactsInManifest.proj(65,5): error : Sequence contains more than one element
   at System.Linq.ThrowHelper.ThrowMoreThanOneElementException()
   at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source)
   at Microsoft.DotNet.Build.Tasks.Feed.PublishArtifactsInManifest.PublishBlobsToAzureStorageNugetFeedAsync(List`1 blobsToPublish, IMaestroApi client, Build buildInformation, FeedConfig feedConfig)
   at Microsoft.DotNet.Build.Tasks.Feed.PublishArtifactsInManifest.HandleBlobPublishingAsync(IMaestroApi client, Build buildInformation) in /_/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifest.cs:line 497
   at Microsoft.DotNet.Build.Tasks.Feed.PublishArtifactsInManifest.ExecuteAsync() in /_/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifest.cs:line 199

Build FAILED.
```

Exactly what does that exception mean? For example is this failure a statement that we need to remove duplication between our manifest files (in contrast to conversations we've had @riarenas @JohnTortugo) ☹️

JohnTortugo commented:
I think the problem here was that same blobs (or same blob name) was described in two different manifests. I'll take a closer look in a few mins.

dougbu commented:
Thought we were pretty sure the duplication between (say) aspnetcore-Linux_x64.xml and aspnetcore-Linux_x64-signed.xml was fine?

I do see some differences in the `<Build />` attributes in those files. Does any of the following matter?
- `InitialAssetsLocation` is '/_apis/build/builds//artifacts' in aspnetcore-Linux_x64.xml but 'https://dev.azure.com/dnceng/internal/_apis/build/builds/347936/artifacts' in aspnetcore-Linux_x64-signed.xml
- `AzureDevOpsProject` is empty in aspnetcore-Linux_x64.xml but 'internal' in aspnetcore-Linux_x64-signed.xml
- `AzureDevOpsBuildId` is empty in aspnetcore-Linux_x64.xml but '347936' in aspnetcore-Linux_x64-signed.xml
- `AzureDevOpsBuildDefinitionId` is empty in aspnetcore-Linux_x64.xml but '21' in aspnetcore-Linux_x64-signed.xml
- `AzureDevOpsAccount` is empty in aspnetcore-Linux_x64.xml but 'dncent' in aspnetcore-Linux_x64-signed.xml

(I don't see why there are empty values for any of the above. Thought such values would be set consistently across all build steps.)

dougbu committed:
Do not publish default items from Linux and macOS builds
- handled separately in later signing builds (on Windows)

dougbu committed:
!fixup! Correct name of parameter passed to publish-build-assets.yml template

dougbu commented:
https://dev.azure.com/dnceng/internal/_build/results?buildId=348634&view=results succeeded with the previous commit 🍾 🚀 🍾 Change in last commit won't impact the build itself (just the BAR). So, we're ready to merge once PR validation completes&hellip;
